bin/p2-replicate/main.go

```diff
@@ -77,19 +71,67 @@ func main() {
 		log.Fatalf("Invalid manifest: %s", err)
 	}
 
-	for _, host := range *hosts {
-		_, _, err := store.Pod(kp.RealityPath(host, "p2-preparer"))
-		if err != nil {
-			log.Fatalf("p2 is not running on host %s: %s", host, err)
-		}
+	healthResults, err := healthChecker.Service(manifest.ID())
+	if err != nil {
+		log.Fatalf("Could not get initial health results: %s", err)
+	}
+	order, err := replication.NewMaxOrder(*hosts, len(*hosts)-*minNodes, healthResults)
+	if err != nil {
+		log.Fatalf("Could not instantiate rollout order", err)
+	}
+	repl := replication.Replicator{
+		Manifest: *manifest,
+		Store:    store,
+		Health:   healthChecker,
+		Order:    order,
+		Logger: logging.NewLogger(logrus.Fields{
+			"pod": manifest.ID(),
+		}),
+		Threshold: health.HealthState(*threshold),
+	}
+	repl.Logger.Logger.Formatter = &logrus.TextFormatter{
+		DisableTimestamp: false,
+		FullTimestamp:    true,
+		TimestampFormat:  "15:04:05.000",
 	}
 
-	allocated := allocation.NewAllocation(*hosts...)
+	if err := repl.CheckPreparers(); err != nil {
+		log.Fatalf("Preparer check failed: %s", err)
+	}
 
-	replicator := replication.NewReplicator(*manifest, allocated)
-	replicator.Logger.Logger.Formatter = new(logrus.TextFormatter)
-	replicator.MinimumNodes = *minNodes
+	// create a lock with a meaningful name and set up a renewal loop for it
+	thisHost, err := os.Hostname()
+	if err != nil {
+		log.Fatalf("Could not retrieve hostname: %s", err)
+	}
+	thisUser, err := user.Current()
+	if err != nil {
+		log.Fatalf("Could not retrieve user: %s", err)
+	}
+	lock, err := store.NewLock(fmt.Sprintf("%q from %q at %q", thisUser.Username, thisHost, time.Now()))
+	if err != nil {
+		log.Fatalf("Could not generate lock: %s", err)
+	}
+	defer lock.Destroy()
+	go func() {
+		for range time.Tick(10 * time.Second) {
+			err := lock.Renew()
```

Comment:
renew should return an error since the target session no longer exists, which would kill main. But the lock destruction is deferred on main, so it would only be encountered on the way out of repl.Enact.

There are definitely some questions about destroying the lock in the case of a nonstandard exit where main does not terminate, eg a ctrl-C (or actually, even an os.Exit). I'm going to poke around that some more today.